### PR TITLE
`<optional>`: Fix exception specification for `optional`'s transform-construction

### LINKED ISSUE #3667: <optional>: Throwing transformers will cause the program to terminate
In the following example, the catch clause is unreacheable. Is this a bug?
The program will terminate directly as <code>optional(_Construct_from_invoke_result_tag, _Fn&& _Func, _Ux&& _Arg)</code>'s noexcept doesn't care about whether _Func itself throws. (see also <code>_Optional_destruct_base(_Construct_from_invoke_result_tag, _Fn&& _Func, _Ux&& _Arg)</code>)
```
#include<optional>

int main() {
	std::optional opt(1);
	try {
		opt.transform([](int) {
			throw(0); return 0;
		});
	}
	catch (...) {
		//unreachable.
	}
}
```
In otional::transform:
```
if (this->_Has_value) {
	return optional<_Uty>{_Construct_from_invoke_result_tag{}, _STD forward<_Fn>(_Func), this->_Value};
} else {
	return optional<_Uty>{};
}
```
In the constructor:
```
template <class _Fn, class _Ux>
constexpr optional(_Construct_from_invoke_result_tag _Tag, _Fn&& _Func, _Ux&& _Arg) noexcept(
        is_nothrow_constructible_v<_Ty, invoke_result_t<_Fn, _Ux>>)
: _Mybase(_Tag, _STD forward<_Fn>(_Func), _STD forward<_Ux>(_Arg)) {}
```

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Fixes #3667.

### CODE REVIEW

File: stl/inc/optional [hunk 1]
```diff
@@ -85,7 +106,7 @@ struct _Optional_destruct_base { // either contains a value of _Ty or is empty (
 #if _HAS_CXX23
     template <class _Fn, class _Ux>
     constexpr _Optional_destruct_base(_Construct_from_invoke_result_tag, _Fn&& _Func, _Ux&& _Arg) noexcept(
-        is_nothrow_constructible_v<_Ty, invoke_result_t<_Fn, _Ux>>)
+        _Is_nothrow_constructible_from_invocation<_Ty, _Fn, _Ux>)
```
Comment (cpplearner) [inline stl/inc/optional:109]: `expected` uses
```c++
noexcept(static_cast<_Ty>(_STD invoke(_STD forward<_Fn>(_Func), _STD forward<_Ux>(_Arg))))
```
Should `optional` use the same pattern?

**Review by strega-nil-ms — APPROVED** [review]
Oh wow, we totally missed that. Thanks for the fix, @frederick-vs-ja!

**Review by cpplearner — APPROVED** [review]

File: tests/std/tests/P0798R8_monadic_operations_for_std_optional/test.cpp [hunk 2]
```diff
@@ -109,7 +110,23 @@ constexpr bool test() {
     return true;
 }
 
+void test_gh3667() {
+    class unique_exception : public exception {};
+
+    try {
+        optional<int> opt(1);
+        opt.transform([](int) -> int { throw unique_exception{}; });
```
Comment (StephanTLavavej) [inline tests/std/tests/P0798R8_monadic_operations_for_std_optional/test.cpp:118]: There are two codepaths being altered in the product (non-trivial destructor and trivial destructor), so we should really test both codepaths.

File: tests/std/tests/P0798R8_monadic_operations_for_std_optional/test.cpp [hunk 3]
```diff
@@ -109,7 +110,23 @@ constexpr bool test() {
     return true;
 }
 
+void test_gh3667() {
```
Comment (StephanTLavavej) [inline tests/std/tests/P0798R8_monadic_operations_for_std_optional/test.cpp:113]: Nitpicks: We usually follow the form `test_gh_NNNN` with an extra underscore, and we usually include a comment reminding us what the bug title was.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! FYI @strega-nil-ms I pushed an expansion of test coverage after you approved.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for fixing this bug in new code! :lady_beetle: :hammer_and_wrench: :heart_eyes_cat:

**Pull request merged**